File: test_liste_fiyat.py
# -*- coding: utf-8 -*-
"""discountlessPrice -> liste_fiyat ve ilan_indirim_gecmisi testi.
scraper.py'yi dogrudan import eder, kopya mantik degil.
Kullanim: py test_liste_fiyat.py
"""
import importlib.util
import json
import os
import sys
import tempfile

_BASE = os.path.dirname(os.path.abspath(__file__))
sys.stdout.reconfigure(encoding="utf-8", errors="replace")

spec = importlib.util.spec_from_file_location("scr", os.path.join(_BASE, "scraper.py"))
scr = importlib.util.module_from_spec(spec)
spec.loader.exec_module(scr)

gecti = basarisiz = 0


def ok(ad, kosul, detay=""):
    global gecti, basarisiz
    if kosul:
        gecti += 1
        print("  PASS  " + ad)
    else:
        basarisiz += 1
        print("  FAIL  " + ad + ("  -> " + str(detay) if detay else ""))


def depot(market, price, **ek):
    d = {"marketAdi": market, "price": price}
    d.update(ek)
    return d


def urun(depots, ad="Test Urun 1 Kg"):
    return {"title": ad, "main_category": "Test", "refinedVolumeOrWeight": "1 KG",
            "imageUrl": None, "productDepotInfoList": depots}


print("\n=== 1. parse_product: liste_fiyat ===")
p = scr.parse_product(urun([depot("migros", 35.18, discountlessPrice=46.9)]), "Test", "test")
mf = p["market_fiyatlari"][0]
ok("discountlessPrice > price -> liste_fiyat eklendi", mf.get("liste_fiyat") == 46.9, mf)
ok("  market ve fiyat degismedi", mf.get("market") == "migros" and mf.get("fiyat") == 35.18, mf)

p = scr.parse_product(urun([depot("bim", 50.0, discountlessPrice=50.0)]), "Test", "test")
mf = p["market_fiyatlari"][0]
ok("discountlessPrice == price -> liste_fiyat YOK", "liste_fiyat" not in mf, mf)

p = scr.parse_product(urun([depot("bim", 50.0, discountlessPrice=40.0)]), "Test", "test")
mf = p["market_fiyatlari"][0]
ok("discountlessPrice < price -> liste_fiyat YOK", "liste_fiyat" not in mf, mf)

p = scr.parse_product(urun([depot("a101", 12.5, discountlessPrice=None)]), "Test", "test")
mf = p["market_fiyatlari"][0]
ok("discountlessPrice None -> liste_fiyat YOK", "liste_fiyat" not in mf, mf)

p = scr.parse_product(urun([depot("sok", 9.9)]), "Test", "test")
mf = p["market_fiyatlari"][0]
ok("alan hic yoksa -> liste_fiyat YOK", "liste_fiyat" not in mf, mf)
ok("  indirimsiz kayit HALA tam olarak {market,fiyat}", set(mf.keys()) == {"market", "fiyat"}, mf)

print("\n=== 2. MEVCUT YAPI BOZULMADI ===")
p = scr.parse_product(urun([depot("migros", 35.18, discountlessPrice=46.9),
                            depot("sok", 40.0)]), "Test", "test")
ok("en_dusuk_fiyat hala SATIS fiyatindan (liste fiyattan degil)", p["en_dusuk_fiyat"] == 35.18, p["en_dusuk_fiyat"])
ok("market_fiyatlari uzunlugu degismedi", len(p["market_fiyatlari"]) == 2, p["market_fiyatlari"])
ok("urun ust seviye anahtarlari degismedi",
   set(p.keys()) == {"_sid", "ad", "ana_kategori", "agirlik_hacim", "resim", "en_dusuk_fiyat", "market_fiyatlari"},
   sorted(p.keys()))
ok("fiyati None olan depot hala atlaniyor",
   len(scr.parse_product(urun([depot("x", None, discountlessPrice=99.0)]), "T", "t")["market_fiyatlari"]) == 0)

print("\n=== 3. ilan_indirim_gecmisi ===")
fn = getattr(scr, "_apply_ilan_indirim_gecmisi", None)
ok("_apply_ilan_indirim_gecmisi tanimli", fn is not None)

if fn:
    def calistir(yeni, eski_json=None):
        tf = tempfile.NamedTemporaryFile("w", suffix=".json", delete=False, encoding="utf-8")
        if eski_json is not None:
            json.dump(eski_json, tf, ensure_ascii=False)
        tf.close()
        yol = tf.name if eski_json is not None else os.path.join(tempfile.gettempdir(), "yok_boyle_dosya.json")
        fn(yeni, yol)
        try:
            os.unlink(tf.name)
        except OSError:
            pass
        return yeni

    u1 = {"_sid": "s1", "market_fiyatlari": [{"market": "migros", "fiyat": 35.18, "liste_fiyat": 46.9}]}
    calistir([u1])
    g = u1.get("ilan_indirim_gecmisi")
    ok("ilan varken kayit yazildi", isinstance(g, list) and len(g) == 1, g)
    if g:
        k = g[0]
        ok("  kayit alanlari {tarih,market,liste_fiyat,satis_fiyat}",
           set(k.keys()) == {"tarih", "market", "liste_fiyat", "satis_fiyat"}, k)
        ok("  degerler dogru", k["market"] == "migros" and k["liste_fiyat"] == 46.9 and k["satis_fiyat"] == 35.18, k)

    u2 = {"_sid": "s2", "market_fiyatlari": [{"market": "sok", "fiyat": 10.0}]}
    calistir([u2])
    ok("ilan YOKKEN kayit yazilmiyor", u2.get("ilan_indirim_gecmisi") == [], u2.get("ilan_indirim_gecmisi"))

    # ayni deger tekrar -> yeni kayit yok
    eski = [{"_sid": "s3", "ilan_indirim_gecmisi": [
        {"tarih": "2020-01-01", "market": "migros", "liste_fiyat": 46.9, "satis_fiyat": 35.18}]}]
    u3 = {"_sid": "s3", "market_fiyatlari": [{"market": "migros", "fiyat": 35.18, "liste_fiyat": 46.9}]}
    calistir([u3], eski)
    ok("ayni ilan tekrar -> yeni kayit YOK", len(u3["ilan_indirim_gecmisi"]) == 1, u3["ilan_indirim_gecmisi"])

    # deger degisti -> yeni kayit
    u4 = {"_sid": "s3", "market_fiyatlari": [{"market": "migros", "fiyat": 29.9, "liste_fiyat": 46.9}]}
    calistir([u4], eski)
    ok("ilan degisti -> yeni kayit eklendi", len(u4["ilan_indirim_gecmisi"]) == 2, u4["ilan_indirim_gecmisi"])

    # farkli market ayri izleniyor
    u5 = {"_sid": "s3", "market_fiyatlari": [
        {"market": "migros", "fiyat": 35.18, "liste_fiyat": 46.9},
        {"market": "carrefour", "fiyat": 33.0, "liste_fiyat": 44.0}]}
    calistir([u5], eski)
    ok("farkli market ayri kayit acti", len(u5["ilan_indirim_gecmisi"]) == 2, u5["ilan_indirim_gecmisi"])

print("\n=== 4. scrape_category akisi cagiriyor mu ===")
src = open(os.path.join(_BASE, "scraper.py"), encoding="utf-8").read()
ok("_apply_ilan_indirim_gecmisi scrape akisinda cagriliyor",
   src.count("_apply_ilan_indirim_gecmisi(") >= 2, src.count("_apply_ilan_indirim_gecmisi("))
ok("discountlessPrice parse_product icinde okunuyor",
   "discountlessPrice" in src.split("def parse_product")[1].split("def ")[0])

print("\nPASS=%d  FAIL=%d" % (gecti, basarisiz))
sys.exit(1 if basarisiz else 0)

File: test_liste_fiyat.mjs
// Urun detayinda liste_fiyat gosterimi testi.
// app.js'ten fonksiyon KAYNAGINI cikarip vm'de calistirir -- kopya mantik degil.
import fs from 'fs';
import vm from 'vm';

const APP = fs.readFileSync('app.js', 'utf8');
const CSS = fs.readFileSync('style.css', 'utf8');

let pass = 0, fail = 0;
const ok = (ad, kosul, detay = '') => {
  if (kosul) { pass++; console.log('  PASS  ' + ad); }
  else { fail++; console.log('  FAIL  ' + ad + (detay ? '  -> ' + detay : '')); }
};

function fnKaynak(ad) {
  let bas = APP.indexOf('function ' + ad + '(');
  if (bas < 0) return null;
  if (APP.slice(Math.max(0, bas - 6), bas) === 'async ') bas -= 6;
  let derinlik = 0;
  for (let j = APP.indexOf('{', bas); j < APP.length; j++) {
    const c = APP[j];
    if (c === '{') derinlik++;
    else if (c === '}') { derinlik--; if (derinlik === 0) return APP.slice(bas, j + 1); }
  }
  return null;
}

console.log('\n=== 0. YAPI ===');
const varFn = !!fnKaynak('listeFiyatHTML');
ok('function listeFiyatHTML tanimli', varFn);

function kur() {
  const ctx = {
    console,
    tlHTML: v => '<span class="fp">' + String(v).replace('.', ',') + ' ₺</span>',
    tl: v => String(v).replace('.', ',') + ' ₺',
  };
  vm.createContext(ctx);
  vm.runInContext(fnKaynak('listeFiyatHTML'), ctx);
  return ctx;
}
const calis = (ctx, i) => vm.runInContext(i, ctx);

if (varFn) {
  const ctx = kur();
  console.log('\n=== 1. liste_fiyat VARKEN ===');
  const h = calis(ctx, 'listeFiyatHTML({market:"migros", fiyat:35.18, liste_fiyat:46.9})');
  ok('bos degil', !!h, JSON.stringify(h));
  ok('liste fiyati yaziliyor', /46,9/.test(h), h);
  ok('indirim yuzdesi yaziliyor (%25)', /%25/.test(h), h);
  ok('kendi sinifini kullaniyor (.detay-mkt-liste)', /detay-mkt-liste/.test(h), h);

  console.log('\n=== 2. liste_fiyat YOKKEN / GECERSIZKEN -> bos string ===');
  ok('liste_fiyat yok', calis(ctx, 'listeFiyatHTML({market:"sok", fiyat:9.9})') === '');
  ok('liste_fiyat null', calis(ctx, 'listeFiyatHTML({market:"sok", fiyat:9.9, liste_fiyat:null})') === '');
  ok('liste_fiyat == fiyat', calis(ctx, 'listeFiyatHTML({market:"sok", fiyat:9.9, liste_fiyat:9.9})') === '');
  ok('liste_fiyat < fiyat', calis(ctx, 'listeFiyatHTML({market:"sok", fiyat:9.9, liste_fiyat:5})') === '');
  ok('fiyat null', calis(ctx, 'listeFiyatHTML({market:"sok", fiyat:null, liste_fiyat:20})') === '');
  ok('arguman yok', calis(ctx, 'listeFiyatHTML(null)') === '');

  // 2026-09-13 olcumu: ilan tasiyan 2.776 kaydin 749'u (%27) bizim o markette
  // HIC gormedigimiz bir fiyat (kaynak da discount=false diyor). Ornek:
  // Finish Quantum ilan 858,00 iken 103 gunde gordugumuz en yuksek 849,90.
  console.log('\n=== 1b. GOZLENMEMIS ILAN: yuzde rozeti BASILMIYOR ===');
  {
    const h2 = calis(ctx, 'listeFiyatHTML({market:"carrefour", fiyat:399.9, liste_fiyat:858}, 849.9)');
    ok('ilan YINE gosteriliyor (marketin beyani silinmiyor)', /858/.test(h2), h2);
    ok('  "-%N" rozeti YOK (biz dogrulamadik)', !/%\d/.test(h2), h2);
    ok('  yerine "biz gormedik" notu var', /görmedik/.test(h2), h2);
    const h3 = calis(ctx, 'listeFiyatHTML({market:"carrefour", fiyat:399.9, liste_fiyat:858}, 858)');
    ok('kontrol: gozlem DESTEKLIYORSA rozet yine basiliyor', /%53/.test(h3), h3);
    const h4 = calis(ctx, 'listeFiyatHTML({market:"c", fiyat:100, liste_fiyat:200})');
    ok('kontrol: gozlem bilinmiyorsa bugunku davranis korunuyor', /%50/.test(h4), h4);
  }

  console.log('\n=== 3. YUZDE HESABI ===');
  const y = (l, f) => (calis(ctx, `listeFiyatHTML({market:"m", fiyat:${f}, liste_fiyat:${l}})`).match(/%(\d+)/) || [])[1];
  ok('185.9 -> 129.9  = %30', y(185.9, 129.9) === '30', y(185.9, 129.9));
  ok('199 -> 179      = %10', y(199, 179) === '10', y(199, 179));
  ok('20 -> 15.9      = %21', y(20, 15.9) === '21', y(20, 15.9));
}

console.log('\n=== 4. DETAY SATIRINA BAGLANMIS MI ===');
{
  const det = APP.slice(APP.indexOf('function openDetay('), APP.indexOf('function openDetay(') + 3800);
  ok('detay market satirlari TEK yerden uretiliyor', /_detayMarketSatirlariHTML\(/.test(det), '');
  // Uretici openDetay'in DISINDA: uc test (al_zamani, esit_fiyat, supheli) o
  // fonksiyonu SABIT karakter penceresiyle kesip icinde cagri ariyor; govdeye
  // eklenen her satir aranan cagrilari disari itiyor (CLAUDE.md'de kayitli
  // tuzak, 2026-09-13'te bir kez daha yasandi: cagrilar 4299-4619, pencere 4000).
  const satir = fnKaynak('_detayMarketSatirlariHTML') || '';
  ok('  uretici listeFiyatHTML cagiriyor', /listeFiyatHTML\s*\(/.test(satir), '');
  ok('  detay-mkt-row icinde cagriliyor', /detay-mkt-row[\s\S]{0,400}listeFiyatHTML/.test(satir), '');
  ok('  gozlenen en yuksek fiyat ikinci arguman olarak veriliyor',
     /listeFiyatHTML\(f,\s*gozlenen/.test(satir), '');
  ok('  gozlem haritasi _marketEnYuksekHaritasi\'ndan geliyor',
     /_marketEnYuksekHaritasi\(/.test(satir), '');
  ok('  fiyatin KAYNAGI (magaza/il/saat) satirin altinda basiliyor',
     /_fiyatKaynagiHTML\(f\)/.test(satir), '');
}

console.log('\n=== 4c. GOZLENEN EN YUKSEK FIYAT HARITASI ===');
{
  const ctx4 = { _gecmisCache: null };
  vm.createContext(ctx4);
  vm.runInContext(fnKaynak('_marketEnYuksekHaritasi') || 'function _marketEnYuksekHaritasi(){}', ctx4);
  const c = i => vm.runInContext(i, ctx4);
  ok('gecmis INMEDIYSE null (bilmedigimizi iddiaya cevirmiyoruz)', c('_marketEnYuksekHaritasi("x")') === null);
  vm.runInContext('_gecmisCache = { x: [{t:"2026-09-01",m:"carrefour",f:849.9},{t:"2026-09-07",m:"carrefour",f:399.9},{t:"2026-09-01",m:"bim",f:"120"}] };', ctx4);
  const h = c('_marketEnYuksekHaritasi("x")');
  ok('  market bazinda EN YUKSEK gozlem', h && h.carrefour === 849.9, JSON.stringify(h));
  ok('  dize fiyat sayiya cevriliyor (alfabetik karsilastirma tuzagi)', h && h.bim === 120, JSON.stringify(h));
  ok('  kaydi olmayan urun icin BOS harita (null degil)',
     JSON.stringify(c('_marketEnYuksekHaritasi("yok")')) === '{}', JSON.stringify(c('_marketEnYuksekHaritasi("yok")')));
}

console.log('\n=== 4b. FIYAT KAYNAGI SATIRI ===');
{
  const ctx3 = { _kacir: s => String(s).replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;'),
                 ZAM_AYLAR: ['Ocak','Şubat','Mart','Nisan','Mayıs','Haziran','Temmuz','Ağustos','Eylül','Ekim','Kasım','Aralık'] };
  vm.createContext(ctx3);
  vm.runInContext(fnKaynak('_fiyatZamaniYazi') + '\n' + fnKaynak('_fiyatKaynagiHTML'), ctx3);
  const c = i => vm.runInContext(i, ctx3);
  const h = c('_fiyatKaynagiHTML({market:"carrefour", fiyat:399.9, depot_ad:"Istanbul Üsküdar Salacak Mını", depot_il:"İstanbul", fiyat_guncelleme:"2026-09-13T08:50"})');
  ok('magaza adi yaziliyor', /Üsküdar Salacak/.test(h), h);
  ok('  guncelleme zamani okunur bicimde', /13 Eylül 08:50/.test(h), h);
  ok('  il magaza adinda zaten geciyorsa TEKRAR yazilmiyor', (h.match(/İstanbul/g) || []).length === 1, h);
  const h2 = c('_fiyatKaynagiHTML({market:"migros", fiyat:1, depot_ad:"Pelitlik Cd M", depot_il:"Antalya"})');
  ok('il magaza adinda yoksa EKLENIYOR', /Antalya/.test(h2), h2);
  ok('veri yoksa satir HIC cizilmiyor', c('_fiyatKaynagiHTML({market:"bim", fiyat:1})') === '');
  ok('kacis uygulaniyor', /&lt;img/.test(c('_fiyatKaynagiHTML({market:"bim", fiyat:1, depot_ad:"<img src=x>"})')));
}

console.log('\n=== 5. KART TARAFINA DOKUNULMADI ===');
{
  const kart = fnKaynak('cardHTML') || '';
  ok('cardHTML listeFiyatHTML cagirmiyor', !/listeFiyatHTML/.test(kart));
  ok('cardHTML liste_fiyat okumuyor', !/liste_fiyat/.test(kart));
  const strip = fnKaynak('_stripKartHTML') || '';
  ok('_stripKartHTML liste_fiyat okumuyor', !/liste_fiyat/.test(strip));
  const firsat = fnKaynak('_firsatKartHtml') || '';
  ok('_firsatKartHtml liste_fiyat okumuyor', !/liste_fiyat/.test(firsat));
}

console.log('\n=== 6. MEVCUT ROZET SISTEMI DEGISMEDI ===');
{
  ok('urunRozetleriHTML duruyor', /function urunRozetleriHTML/.test(APP));
  ok('supheliRozetHTML duruyor', /function supheliRozetHTML/.test(APP));
  ok('gercekIndirimRozetiHesapla duruyor', /function gercekIndirimRozetiHesapla/.test(APP));
  ok('yeni rozet sinifi UYDURULMADI', !/indirim-rozet\s+ilan|ilan-rozet/.test(APP));
}

console.log('\n=== 7. MEVCUT OKUYUCULAR KIRILMADI (fiyatlariTemizle nesneyi koruyor) ===');
{
  const ctx2 = {};
  vm.createContext(ctx2);
  vm.runInContext(fnKaynak('fiyatlariTemizle'), ctx2);
  const girdi = [{ market: 'migros', fiyat: 35.18, liste_fiyat: 46.9 }, { market: 'sok', fiyat: 40 }];
  const r = vm.runInContext('fiyatlariTemizle(' + JSON.stringify(girdi) + ')', ctx2);
  ok('gecerli uzunlugu 2', r.gecerli.length === 2, JSON.stringify(r));
  ok('liste_fiyat filtreden gecti (nesne yeniden kurulmuyor)', r.gecerli[0].liste_fiyat === 46.9, JSON.stringify(r.gecerli[0]));
  ok('liste_fiyat OLMAYAN kayit hala saglam', r.gecerli[1].market === 'sok' && r.gecerli[1].fiyat === 40);
}

console.log('\n=== 8. CSS ===');
{
  // alt secicileri de yakala: ".detay-mkt-liste s { ... }"
  const k = (CSS.match(/[^\n{}]*\.detay-mkt-liste[^{}]*\{[^}]*\}/g) || []).join('\n');
  ok('.detay-mkt-liste kurali var', k.length > 20, 'uzunluk=' + k.length);
  ok('ustu cizili (line-through)', /line-through/.test(k), k);
  ok('KIRMIZI kullanilmamis', !/#(DC2626|dc2626|EF4444|ef4444)/.test(k), k);
}

console.log('\nPASS=' + pass + '  FAIL=' + fail);
process.exit(fail ? 1 : 0);
